File: map-gen-script/src/map_generator/topologies/v_shape.py
import random
from .base_topology import BaseTopology
import copy
from src.map_generator.models.path_info import PathInfo, Coord
from src.utils.geometry import add_vectors, FORWARD_X, FORWARD_Z, BACKWARD_X

class VShapeTopology(BaseTopology):
    """
    Tạo ra một con đường hình chữ V trên mặt phẳng 2D.
    Lý tưởng cho các bài học về tuần tự hoặc hàm đơn giản.
    """
    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể hình chữ V với độ dài cánh tay khác nhau.
        """
        count = 0
        # Lặp qua các độ dài cánh tay có thể có
        for arm_len in range(2, 7):
            if count >= max_variants: return
            
            variant_params = copy.deepcopy(params)
            variant_params['arm_length'] = arm_len
            yield self.generate_path_info(grid_size, variant_params)
            count += 1

    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        """
        Tạo ra một đường đi hình chữ V.

        Args:
            params (dict):
                - arm_length (int): Số cặp bước đi trong mỗi nhánh chữ V.

        Returns:
            PathInfo: Một đối tượng chứa thông tin về đường đi.
        """
        print("    LOG: Generating 'v_shape' topology...")

        arm_len = params.get('arm_length', random.randint(3, 5))
        y = 0

        # [SỬA LỖI] Tính toán kích thước cần thiết dựa trên đường đi zigzag
        # Mỗi bước trong arm_len sẽ tạo ra 2 ô (1 ngang, 1 dọc)
        required_width = arm_len * 2
        required_depth = arm_len * 2
        
        # Đảm bảo nằm gọn trong map
        if required_width > grid_size[0] - 2: required_width = grid_size[0] - 2
        if required_depth > grid_size[2] - 2: required_depth = grid_size[2] - 2
        
        start_x = random.randint(1, grid_size[0] - required_width - 1)
        start_z = random.randint(1, grid_size[2] - required_depth - 1)

        start_pos: Coord = (start_x, y, start_z)
        # Sửa lỗi: Đường đi phải luôn bắt đầu bằng vị trí xuất phát
        path_coords: list[Coord] = [start_pos]
        current_pos = start_pos

        # [SỬA LỖI] Thay thế di chuyển chéo bằng di chuyển zigzag (bậc thang)
        
        # Nhánh đầu tiên của chữ V (đi tiến và sang phải)
        # Hướng di chuyển tổng thể: (+X, +Z)
        for _ in range(arm_len):
            # Bước 1: Đi tiến (trục Z)
            current_pos = add_vectors(current_pos, FORWARD_Z) # (0, 0, 1)
            path_coords.append(current_pos)
            # Bước 2: Đi sang phải (trục X)
            current_pos = add_vectors(current_pos, FORWARD_X) # (1, 0, 0)
            path_coords.append(current_pos)

        # Nhánh thứ hai của chữ V (đi tiến và sang trái)
        # Hướng di chuyển tổng thể: (-X, +Z)
        for _ in range(arm_len):
            # Bước 1: Đi tiến (trục Z)
            current_pos = add_vectors(current_pos, FORWARD_Z) # (0, 0, 1)
            path_coords.append(current_pos)
            # Bước 2: Đi sang trái (trục X)
            current_pos = add_vectors(current_pos, BACKWARD_X) # (-1, 0, 0)
            path_coords.append(current_pos)

        target_pos = path_coords[-1]

        # [NEW] Tạo metadata cho Placers
        # Điểm thấp nhất của V là corner (apex)
        apex_idx = arm_len * 2  # Vị trí đỉnh V trong path
        apex = path_coords[apex_idx]
        seg1 = path_coords[:apex_idx + 1]  # Nhánh trái
        seg2 = path_coords[apex_idx:]       # Nhánh phải
        
        metadata = {
            "segments": [seg1, seg2],
            "corner": apex,
            "arm_length": arm_len,
        }

        # placement_coords chính là path_coords vì mỗi ô đi qua đều cần có đất
        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=path_coords,
            placement_coords=path_coords,
            metadata=metadata
        )

File: apps/map-builder-app/src/map-generator/topologies/VShape.ts
/**
 * V Shape Topology
 * Creates a V-shaped path - ideal for sequential or simple function lessons
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';

export class VShapeTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    const baseArmLength = params.arm_length || 3;
    
    for (let i = 0; i < maxVariants; i++) {
      const armLength = baseArmLength + i;
      if (armLength > 6) break;
      yield this.generatePathInfo(gridSize, { ...params, arm_length: armLength });
    }
  }

  generatePathInfo(gridSize: [number, number, number], params: Record<string, any>): IPathInfo {
    const armLength = params.arm_length || 3;
    const startX = params.start_x || 2;
    const startZ = params.start_z || 2;
    const y = 0;

    const startPos: Coord = [startX, y, startZ];
    const pathCoords: Coord[] = [startPos];
    let current: Coord = [...startPos];

    // First arm of V: move forward (+Z) and right (+X) in zigzag
    for (let i = 0; i < armLength; i++) {
      current = [current[0], current[1], current[2] + 1];
      pathCoords.push([...current]);
      current = [current[0] + 1, current[1], current[2]];
      pathCoords.push([...current]);
    }

    // Second arm of V: move forward (+Z) and left (-X) in zigzag
    for (let i = 0; i < armLength; i++) {
      current = [current[0], current[1], current[2] + 1];
      pathCoords.push([...current]);
      current = [current[0] - 1, current[1], current[2]];
      pathCoords.push([...current]);
    }

    const targetPos = pathCoords[pathCoords.length - 1];
    const apexIdx = armLength * 2;
    const apex = pathCoords[apexIdx];

    return {
      start_pos: startPos,
      target_pos: targetPos,
      path_coords: pathCoords,
      placement_coords: pathCoords.slice(1, -1),
      obstacles: [],
      metadata: {
        topology_type: 'v_shape',
        segments: [pathCoords.slice(0, apexIdx + 1), pathCoords.slice(apexIdx)],
        corner: apex,
        arm_length: armLength,
      },
    };
  }
}
